File: 백준/Silver/15663. N과 M （9）/N과 M （9）.cc
#include <iostream>
#include <algorithm>
using namespace std;
int n,m;
int arr[10], memo[10], isused[10];

void go(int k){
	if(k == m){
		for(int i = 0; i < m; i++) cout << memo[i] << " ";
		cout << "\n";
		return;
	}

	int tmp = 0;
	for(int i = 0; i < n; i++){
		if(arr[i] != tmp && !isused[i]){
			memo[k] = arr[i];
			isused[i] = 1;
			go(k + 1);
			tmp = arr[i];
			isused[i] = 0;
		}
	}
	
}

int main() {
	cin >> n >> m;
	for(int i = 0; i < n; i++) cin >> arr[i];
	sort(arr,arr + n);
	go(0);
	return 0;
}

File: 백준/Silver/15663. N과 M （9）/N과 M （9）.java
import java.io.*;
import java.util.*;

class Main {
	public static int[] arr,c;
	public static int n, m;
	public static Set<String> set = new LinkedHashSet<>();
	public static StringBuffer sb = new StringBuffer();
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st1 = new StringTokenizer(br.readLine());
		StringTokenizer st2 = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st1.nextToken());
		m = Integer.parseInt(st1.nextToken());
		arr = new int[n];
		c = new int[n];
		for(int i = 0; i < n; i++){
			arr[i] = Integer.parseInt(st2.nextToken());
		}
		Arrays.sort(arr);
		go(0,"");
		for(String str : set){
			sb.append(str + "\n");
		}
		
		System.out.print(sb.toString());
	}
	
	public static void go(int k, String tmp){
		if(k == m){
			set.add(tmp.trim());
			return;
		}
		
		for(int i = 0; i < n; i++){
			if(c[i] == 1) continue;
			c[i] = 1;
			go(k + 1, tmp + " " + arr[i]);
			c[i] = 0;
		}
	}
}